### PR TITLE
[POAE-2309] [Velox To Substrait] Convert Velox HashJoinNode to Substrait JoinRel

### PR DESCRIPTION
**Pull request by chaojun-zhang**
Convert Velox HashJoinNode to Substrait JoinRel

### CODE REVIEW

File: velox/substrait/SubstraitToVeloxPlan.h [hunk 1]
```diff
@@ -132,6 +138,16 @@ class SubstraitVeloxPlanConverter {
       const ::substrait::Expression& substraitFilter,
       std::vector<::substrait::Expression_ScalarFunction>& scalarFunctions);
 
+  /// Extract join keys from joinExpression.
+  /// joinExpression is a boolean condition that describes whether each record
+  /// from the left set �match� the record from the right set. The condition
```
Comment (winningsix) [inline velox/substrait/SubstraitToVeloxPlan.h]: b/c Chinese characters?

Comment (chaojun-zhang) [reply velox/substrait/SubstraitToVeloxPlan.h]: updated

File: velox/substrait/VeloxToSubstraitPlan.cpp [hunk 2]
```diff
@@ -378,57 +350,87 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
...
+  if (joinNode->joinType() == core::JoinType::kLeftSemi) {
+    joinOutputRowType = joinSources[0]->outputType();
+  } else if (joinNode->joinType() == core::JoinType::kRightSemi) {
+    joinOutputRowType = joinSources[1]->outputType();
+  }
+  auto joinOutputTypeSize = joinOutputRowType->size();
+
+  // Insert a project rel for outputRowType of HashJoinNode.
+  auto projectRel = rel->mutable_project();
+  auto projectEmitRel = projectRel->mutable_common()->mutable_emit();
```
Comment (jikunshang) [inline velox/substrait/VeloxToSubstraitPlan.cpp:370]: what's this for?

File: velox/substrait/VeloxToSubstraitPlan.cpp [hunk 3]
```diff
@@ -378,57 +350,87 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
...
-    const std::shared_ptr<const core::AbstractJoinNode>& joinNode,
-    ::substrait::JoinRel* joinRel) {
-  std::vector<core::PlanNodePtr> sources = joinNode->sources();
+    const std::shared_ptr<const core::HashJoinNode> joinNode,
+    ::substrait::Rel* rel) {
+  const auto& joinSources = joinNode->sources();
+  auto joinOutputRowType =
+      joinSources[0]->outputType()->unionWith(joinSources[1]->outputType());
+
+  if (joinNode->joinType() == core::JoinType::kLeftSemi) {
```
Comment (jikunshang) [inline velox/substrait/VeloxToSubstraitPlan.cpp:361]: for left semi join, `joinSources[1]->outputType()` should be empty?

File: velox/substrait/VeloxToSubstraitPlan.cpp [hunk 4]
```diff
@@ -378,57 +350,87 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
...
+void VeloxToSubstraitPlanConvertor::toSubstraitJoin(
     google::protobuf::Arena& arena,
-    const std::shared_ptr<const core::AbstractJoinNode>& joinNode,
-    ::substrait::JoinRel* joinRel) {
-  std::vector<core::PlanNodePtr> sources = joinNode->sources();
+    const std::shared_ptr<const core::HashJoinNode> joinNode,
+    ::substrait::Rel* rel) {
+  const auto& joinSources = joinNode->sources();
+  auto joinOutputRowType =
+      joinSources[0]->outputType()->unionWith(joinSources[1]->outputType());
```
Comment (jikunshang) [inline velox/substrait/VeloxToSubstraitPlan.cpp:359]:
```
CREATE TABLE table_left(l_a BIGINT, l_b INTEGER, l_c DOUBLE, l_d FLOAT)
CREATE TABLE table_right(r_a BIGINT, r_b INTEGER, r_c DOUBLE, r_d FLOAT)
SELECT l_a FROM table_left JOIN table_right ON l_a = r_a
```
`joinSources[0]->outputType ` will be all columns or just col_a?

File: velox/substrait/VeloxToSubstraitPlan.cpp (hunk 2 above)
Comment (chaojun-zhang) [reply velox/substrait/VeloxToSubstraitPlan.cpp:370]: 'outputRowType' in velox hashJoinNode can be part of field reference in either left or/and right table. so we have to insert a projectRel for this

File: velox/substrait/VeloxToSubstraitPlan.cpp (hunk 3 above)
Comment (chaojun-zhang) [reply velox/substrait/VeloxToSubstraitPlan.cpp:361]: yes. the output of right table should not be include in join projections if joinType = leftSemi

File: velox/substrait/VeloxToSubstraitPlan.cpp (hunk 4 above)
Comment (chaojun-zhang) [reply velox/substrait/VeloxToSubstraitPlan.cpp:359]: actually outputType should always reference  left table and/or right table according the join type,  but in velox the HashJoinNode  is actually a combination of  project rel + join rel  due to the output layout can be part of columns in either left and/or right table.

File: velox/type/Type.h [hunk 5]
```diff
@@ -835,7 +835,8 @@ class RowType : public TypeBase<TypeKind::ROW> {
   void printChildren(std::stringstream& ss, std::string_view delimiter = ",")
       const;
 
-  std::shared_ptr<RowType> unionWith(std::shared_ptr<RowType>& rowType) const;
+  std::shared_ptr<const RowType> unionWith(
```
Comment (ZJie1) [inline velox/type/Type.h:838]: Using RowTypePtr?

File: velox/substrait/VeloxToSubstraitPlan.cpp (hunk 2 above)
Comment (jikunshang) [reply velox/substrait/VeloxToSubstraitPlan.cpp:370]: only used in Line 377? any other place I miss?

**Review by ZJie1 — APPROVED** [review]

**Pull request merged**